src/shared/metrics.c

```diff
@@ -0,0 +1,234 @@
+        if (r != 0)
+                return r;
+
+        if (!FLAGS_SET(flags, SD_VARLINK_METHOD_MORE))
+                return sd_varlink_error(link, SD_VARLINK_ERROR_EXPECTED_MORE, NULL);
+
+        MetricFamilyContext ctx = {
+                .metric_family = NULL,
+                .link = link,
+                .previous = NULL};
+        for (const MetricFamily *mf = metric_family_table; mf && mf->name; mf++) {
+                assert(mf->iterate_cb);
+
+                ctx.metric_family = mf;
+                r = mf->iterate_cb(&ctx, userdata);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to list metrics for metric family '%s': %m", mf->name);
+        }
+
+        /* produce last metric */
+        if (ctx.previous)
+                return sd_varlink_reply(link, ctx.previous);
+        else
+                return sd_varlink_error(link, SD_VARLINK_ERROR_EXPECTED_MORE, NULL);
+}
+
+static int metric_set_fields(sd_json_variant **v, char **fields) {
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *fields_v = NULL;
+        int r;
+
+        STRV_FOREACH_PAIR(fk, fv, fields) {
+                r = sd_json_variant_merge_objectbo(&fields_v, SD_JSON_BUILD_PAIR_STRING(*fk, *fv));
+                if (r < 0)
+                        return r;
+        }
+
+        return sd_json_variant_set_field(ASSERT_PTR(v), "fields", fields_v);
+}
+
+static int metric_build_send(MetricFamilyContext* context, const char *object, sd_json_variant *value, char **fields) {
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+        int r;
+
+        assert(context);
+        assert(value);
+        assert(context->link);
+        assert(context->metric_family);
+
+        r = sd_json_buildo(
+                        &v,
+                        SD_JSON_BUILD_PAIR_STRING("name", context->metric_family->name),
+                        JSON_BUILD_PAIR_STRING_NON_EMPTY("object", object),
+                        JSON_BUILD_PAIR_VARIANT_NON_NULL("value", value));
+                        /* TODO JSON_BUILD_PAIR_OBJECT_STRV_NOT_NULL */
+        if (r < 0)
+                return r;
+
+        if (fields) { /* NULL => no fields object, empty strv => fields: {} */
+                r = metric_set_fields(&v, fields);
+                if (r < 0)
+                        return r;
+        }
+
+        if (context->previous) {
+                r = sd_varlink_notify(context->link, context->previous);
+                if (r < 0)
+                        return r;
+
+                context->previous = sd_json_variant_unref(context->previous);
+        }
+
+        context->previous = TAKE_PTR(v);
+
+        return 0;
+}
+
+int metric_build_send_string(MetricFamilyContext* context, const char *object, const char *value, char **fields) {
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+
+        int r = sd_json_variant_new_string(&v, value);
```

Comment:
```
int r;
```

`r` should be on a separate line. Always. The maintainers are quite picky about it